byron/ledger/executable-spec/src/Ledger/UTxO/Generators.hs

```diff
@@ -0,0 +1,273 @@
+import Hedgehog.Internal.Tree (NodeT(..), TreeT(..), treeValue)
+import qualified Hedgehog.Internal.Tree as Tree
+
+
+--------------------------------------------------------------------------------
+-- Initial TxOut generator
+--------------------------------------------------------------------------------
+
+-- | Generate a set of initial @TxOut@s from a set of @Addr@s
+genInitialTxOuts :: [Addr] -> Gen [TxOut]
+genInitialTxOuts = Gen.filter (not . null)
+  . genTraverseSubsequence (\a -> TxOut a <$> genLovelace)
+
+genLovelace :: Gen Lovelace
+genLovelace = Lovelace . fromIntegral <$> Gen.word32 (Range.linear 1 10000)
+
+-- | Generate a subsequence of a list of values and traverse the subsequence
+--   with a generator producer
+genTraverseSubsequence :: (a -> Gen b) -> [a] -> Gen [b]
+genTraverseSubsequence genA as =
+  mapGenT (TreeT . (interleaveTreeT . nodeValue =<<) . runTreeT) $ do
+    sub <- Gen.subsequence as
+    traverse (toTreeMaybeT . genA) sub
+
+-- | Generate a list using @genTraverseSubsequence@
+genList :: Range Int -> Gen a -> Gen [a]
+genList range gen = Gen.sized $ \size ->
+  ensure (atLeast $ Range.lowerBound size range) $ genTraverseSubsequence
+    (const gen)
+    (replicate (Range.upperBound size range) ())
+
+-- | Temporarily defined here until hedgehog exposes this function
+interleaveTreeT :: Monad m => [TreeT m a] -> m (NodeT m [a])
+interleaveTreeT = fmap Tree.interleave . traverse runTreeT
+
+
+--------------------------------------------------------------------------------
+-- Tx generator
+--------------------------------------------------------------------------------
+
+-- | Generate a valid transaction from a given @UTxO@
+genTxFromUTxO :: Ord id => id -> [Addr] -> UTxO id -> Gen (Tx id)
+genTxFromUTxO placeholderId addrs utxo =
+  uncurry (Tx placeholderId) <$> Gen.filter
+    (not . null . fst)
+    (genInputOutput
+      (M.keys $ unUTxO utxo)
+      (maybe 0 (unLovelace . value) . flip M.lookup (unUTxO utxo))
+      (fmap (. Lovelace) $ TxOut <$> Gen.element addrs)
+      (unLovelace . value)
+      (\f out -> out { value = Lovelace . f . unLovelace $ value out })
+    )
+
+
+-- | A property to test that the entire shrink tree generated by
+--   @genInputOutput@ maintains the invariant that the inputs and outputs have
+--   equal sums
+propGenInputOutput :: Property
+propGenInputOutput = property $ do
+  ins <- forAll $ Gen.list (Range.linear 0 5) (Gen.integral (Range.linear 0 5))
+  insOutsTree <- forAll $ toTree (genInputOutput ins id (pure id) id id)
+  assert $ all (\(ins', outs) -> sum ins' == sum outs) insOutsTree
+
+-- | Generate a subsequence of the inputs and split the value into a number of
+--   outputs
+--
+--   This is generalised to any type of input where we can view the contained
+--   value, and type of output where we can generate the type from a value, and
+--   view and modify the contained value
+genInputOutput
+  :: [input]
```

Comment:
It'd be nice to have some comments on what the different parameters denote.

Reply:
I forgot you can add comments to the arguments :P